### PR TITLE
Support React 16 ReactDOM.hydrate using a two-pass rendering technique (plus improvements)

### PR DESCRIPTION
**Pull request by albertstill**
- Support React 16 ReactDOM.hydrate using a two-pass rendering technique, this means when values is provided we use values on the client (first pass) to match the server render. This stops the client server mismatch error which can cause wrong markup with [ReactDOM.hydrate](https://github.com/facebook/react/issues/10591). Then in componentDidMount we use the native `window.matchMedia` and re-render if the server guessed the wrong screen dimensions
- Removed `matchmedia` dep as this defaults to window.matchMedia if available
- Improve Babel config
- Added Create React App demo folder 
- Upgrade all deps
- Applied linting to test files

See README.md changes for more info.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
+A dead simple React Higher Order Component (HOC) that uses context for matching media queries.
 
 
 ## Why use this?
 - A simple API which doesnt require you to put `MediaQuery` components all over your code base
 - More performant (you only need 1 parent `MediaQueryProvider` that listens to media events you wish to configure)
 - Easier to test than other react media query libraries
-- Uses [matchmedia](https://github.com/iceddev/matchmedia) for media queries for client and server
-- Abstracted away React context which is experimental (and subject to change for <= React 15)
+- Uses [css-mediaquery](https://github.com/ericf/css-mediaquery) for server side rendering
```
Comment (jooj123) [inline README.md]: "Uses [css-mediaquery](https://github.com/ericf/css-mediaquery) for parsing css media queries that match values (used for server side rendering)"

File: README.md [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
+
+## Why not use this?
+We always recommend using vanilla CSS media queries to build responsive websites, this is simpler and provides a smoother UX, also it mitigates having to guess the screen width during [server side rendering](#server-side-rendering). At Domain we needed to use this component for legacy ad tech and stat reasons and advise against it's use for general responsive website design.
```
Comment (jooj123) [inline README.md]: "We recommend using vanilla CSS media queries to build responsive websites over JS solutions, this is simpler,  provides a smoother UX and also it mitigates having to guess the screen width during [server side rendering](#server-side-rendering). 
Use this library if you need to dramatically alter the page layout between media types (a good example would be if you had to render completely different components)"

File: README.md [hunk 3]
```diff
@@ -68,7 +70,6 @@ const App = (props) => {
 ### `withMedia`
 
 This is a HOC to provide media match props to your component.
-This abstracts away context so that if there is any changes to the API in the future its easier to upgrade (see: [React Context](https://facebook.github.io/react/docs/context.html))
```
Comment (jooj123) [inline README.md]: why was this removed?

File: README.md [hunk 4]
```diff
@@ -95,12 +96,11 @@ export const BaseMyComponent = MyComponent;
 export default withMedia(MyComponent);
 ```
 
-Components wrapped by `withMedia()` won't work with React's usual `ref` mechanism, because the ref supplied will be for `withMedia` rather than the wrapped component. Therefore a prop, `wrappedRef` provides the same function.
+Components wrapped by `withMedia()` won't work with React's usual `ref` mechanism, because the ref supplied will be for `withMedia` rather than the wrapped component. Therefore a prop, `wrappedRef` provides the same function. Note this means the wrapped component can not be a [stateless function](https://github.com/facebook/react/issues/10831).
```
Comment (jooj123) [inline README.md]: "Note: this"

File: README.md [hunk 5]
```diff
@@ -118,6 +118,15 @@ const App = (props) => {
 };
 ```
 
+#### React 16 ReactDOM.hydrate
+
+It's very important to realise a server client mismatch is dangerous when using hydrate in React 16, ReactDOM.hydrate
```
Comment (jooj123) [inline README.md]: 👍

File: package.json [hunk 6]
```diff
@@ -5,9 +5,10 @@
   "main": "dist/index.js",
   "scripts": {
     "prepublish": "npm run build",
-    "build": "mkdir -p dist && babel src --out-dir dist",
-    "test": "mocha test",
-    "lint": "eslint ./src"
+    "build": "NODE_ENV=production babel --out-dir dist src",
```
Comment (jooj123) [inline package.json]: why was this changed? Best to keep this PR related to react 16 compat only

File: package.json [hunk 7]
```diff
@@ -5,9 +5,10 @@
...
+    "test": "NODE_ENV=test mocha test",
```
Comment (jooj123) [inline package.json]: same as above

File: package.json [hunk 8]
```diff
@@ -28,47 +29,45 @@
...
     "react-addons-test-utils": "^15.1.0",
-    "react-dom": "^15.1.0",
-    "sinon": "^3.2.1"
+    "react-dom": "^16.3.2",
+    "sinon": "^5.0.7"
   },
   "dependencies": {
-    "hoist-non-react-statics": "^1.2.0",
-    "matchmedia": "^0.1.2",
+    "babel-runtime": "^6.26.0",
```
Comment (jooj123) [inline package.json]: I dont think this should be a dependancy - see docs:
https://babeljs.io/docs/plugins/transform-runtime/ "The transformation plugin is typically used only in development, but the runtime itself will be depended on by your deployed/published code. See the examples below for more details." - so it depends on use of this library 
We at domain import this separately on server

Can you clarify why its needed with your changes ?

File: package.json [hunk 9]
```diff
@@ -28,47 +29,45 @@
...
+    "css-mediaquery": "^0.1.2",
+    "detect-node": "^2.0.3",
```
Comment (jooj123) [inline package.json]: I think remove - can you create another file called `dom-exists.js`

and have 
```
function domExists() {
  return !!(
    typeof window !== 'undefined'
    && window.document
    && window.document.createElement
  );
}

export default domExists;
```

export default domExists;``` in it - we dont need to worry about another external dep for a few lines of code - and `window` checks should be sufficient

File: package.json [hunk 10]
```diff
@@ -28,47 +29,45 @@
...
   "peerDependencies": {
-    "react": "^15.1.0"
+    "react": "^15.1.0 || ^16.0.0"
   },
   "devDependencies": {
     "babel-cli": "^6.10.1",
-    "babel-preset-es2015": "^6.9.0",
-    "babel-preset-react": "^6.5.0",
-    "babel-preset-stage-0": "^6.5.0",
+    "babel-plugin-transform-es2015-modules-commonjs": "^6.26.2",
```
Comment (jooj123) [inline package.json]: why was this changed ? isnt it easier to just use `babel-preset-es2015` so we dont need to worry about individual babel plugins

File: src/media-query-provider.js [hunk 11]
```diff
@@ -1,68 +1,87 @@
...
+// this is for server side rendering and does not use window.matchMedia
+import cssMediaQuery from 'css-mediaquery';
+import isNode from 'detect-node';
 
 class MediaQueryProvider extends React.Component {
-
   constructor(props) {
     super(props);
+
+    const queryTuples = Object.entries(this.props.queries);
```
Comment (jooj123) [inline src/media-query-provider.js]: Rather than use `Object.entries` can you use:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries#Polyfill
(place this function in another file)

So we dont need to worry about IE support

File: src/media-query-provider.js [hunk 12]
```diff
@@ -1,68 +1,87 @@
...
+
+    const media = queryTuples.reduce((acc, [queryName, query]) => {
+      if (this.props.values) {
+        acc[queryName] = cssMediaQuery.match(query, this.props.values);
+      } else {
+        // if the consumer has not set `values` and is server rendering, default to false
+        // because we don't know the screen size
+        acc[queryName] = isNode ? false : window.matchMedia(query).matches;
```
Comment (jooj123) [inline src/media-query-provider.js]: as mentioned use `domExists` func i supplied in comments to avoid the extra library

File: README.md (hunk 3 above)
Comment (albertstill) [reply README.md]: Didn't think it was relevant now that React 16's Context API is not experimental.

> React previously shipped with an experimental context API

https://reactjs.org/docs/context.html#legacy-api

We should use the legacy one while we support React 15, I'll mention this.

File: package.json (hunk 6 above)
Comment (albertstill) [reply package.json]: Using `babel-preset-react-app` requires that you specify `NODE_ENV`

File: package.json (hunk 8 above)
Comment (albertstill) [reply package.json]: > but the runtime itself will be depended on by your deployed/published code

The transpiled code requires `babel-runtime`, it provides the polyfills so we don't have to worry about the consumers browser version and if they have `babel-polyfill`. It's awesome for libraries, the Babel docs say:

> This plugin is recommended in a library/tool.

File: package.json (hunk 9 above)
Comment (albertstill) [reply package.json]: I added it because the package had loads of downloads and I figured it could possibly be in the consumers bundle anyway. I can remove tho.

File: package.json (hunk 10 above)
Comment (albertstill) [reply package.json]: This was because `babel-preset-react-app` does not transpile CJS modules, because Webpack supports ES modules. But I have actually decided to use the env and react preset instead after taking a closer look at `babel-preset-react-app` src, it's more geared towards an app vs a library e.g it removes proptype warnings for prod.

But yearly presets are not recommended:

> Each yearly preset only compiles what was ratified in that year. babel-preset-env replaces es2015, es2016, es2017 and latest

https://babeljs.io/docs/plugins/

> Without any configuration options, babel-preset-env behaves exactly the same as babel-preset-latest (or babel-preset-es2015, babel-preset-es2016, and babel-preset-es2017 together).

https://babeljs.io/docs/plugins/preset-env/

Also see https://github.com/contra/react-responsive/blob/master/package.json#L77-L80

File: src/media-query-provider.js (hunk 11 above)
Comment (albertstill) [reply src/media-query-provider.js]: Babel's transform runtime has us covered here, see the dist code requires `babel-runtime/core-js/object/entries`, this is more efficient than library maintainers rolling their own polyfills which will cause logic duplication inside the bundles.

File: package.json [hunk 13]
```diff
@@ -5,9 +5,10 @@
   "main": "dist/index.js",
   "scripts": {
     "prepublish": "npm run build",
-    "build": "mkdir -p dist && babel src --out-dir dist",
+    "build": "babel --out-dir dist src",
```
[the remark this answers is not in the document]
Comment (albertstill) [reply package.json]: yes

File: package.json [hunk 14]
```diff
@@ -28,47 +29,46 @@
...
     "react-addons-test-utils": "^15.1.0",
-    "react-dom": "^15.1.0",
-    "sinon": "^3.2.1"
+    "react-dom": "^16.3.2",
+    "sinon": "^5.0.7"
   },
   "dependencies": {
-    "hoist-non-react-statics": "^1.2.0",
-    "matchmedia": "^0.1.2",
+    "babel-plugin-transform-object-rest-spread": "^6.26.0",
```
[the remark this answers is not in the document]
Comment (albertstill) [reply package.json]: yeh fixed it
